File: c/catbrotli.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "brotli/broccoli.h"

void usage() {
    fprintf(stderr,
            "Usage: [-w<window_size>] filename0 filename1 filename2...");
}


int main(int argc, char**argv) {
    unsigned char window_size = 0;
    unsigned char has_window_size = 0;
    unsigned char double_dash = 0;
    size_t buffer_size = 4096;
    int i;
    if (argc == 1) {
        usage();
        return 1;
    }
    for (i = 1; i < argc; ++i) {
        if (strcmp(argv[i], "-help") == 0 && !double_dash) {
            usage();
            return 1;
        }
        if (strcmp(argv[i], "--help") == 0 && !double_dash) {
            usage();
            return 1;
        }
        if (strcmp(argv[i], "-h") == 0 && !double_dash) {
            usage();
            return 1;
        }
        if (strncmp(argv[i], "-w", 2) == 0 && !double_dash) {
            int j;
            has_window_size = 1;
            window_size = atoi(argv[i] + 2);
            for (j = i; j + 1 < argc; ++j) {
                argv[j] = argv[j+1];
            }
            --i;
            --argc;
            continue;
        }
        if (strncmp(argv[i], "-bs", 3) == 0 && !double_dash) {
            int j;
            buffer_size = atoi(argv[i] + 3);
            for (j = i; j + 1 < argc; ++j) {
                argv[j] = argv[j+1];
            }
            --i;
            --argc;
            continue;
        }
        if (strcmp(argv[i], "--") == 0) {
            int j;
            double_dash = 1;
            for (j = i; j + 1 < argc; ++j) {
                argv[j] = argv[j+1];
            }
            --i;
            --argc;
            continue;
        }
    }
    char ** filenames = &argv[1];
    unsigned char * ibuffer = (unsigned char*)malloc(buffer_size);
    unsigned char * obuffer = (unsigned char*)malloc(buffer_size);
    unsigned char* obuffer_ptr = obuffer;
    size_t avail_out = buffer_size;
    struct BroccoliState state;
    int i;
    if (has_window_size) {
        state = BroccoliCreateInstanceWithWindowSize(window_size);
    } else {
        state = BroccoliCreateInstance();
    }
    for (i = 1; i < argc; ++i) {
        BroccoliNewBrotliFile(&state);
        FILE * input_file = fopen(argv[i], "rb");
        if (!input_file) {
            fprintf(stderr, "Could not open %s\n", argv[i]);
            usage();
            return 1;
        }
        while(1) {
            size_t cur_read = fread(ibuffer, 1, buffer_size, input_file);
            const unsigned char *ibuffer_ptr = ibuffer;
            size_t avail_in = cur_read;
            if (cur_read == 0) {
                break;
            }
            while(1) {
                enum BroccoliResult res = BroccoliConcatStream(
                    &state,
                    &avail_in,
                    &ibuffer_ptr,
                    &avail_out,
                    &obuffer_ptr);
                if (res == BroccoliSuccess) {
                    abort();
                }
                if (res == BroccoliNeedsMoreInput) {
                    break;
                }
                if (res == BroccoliNeedsMoreOutput) {
                    fwrite(obuffer, 1, (obuffer_ptr - obuffer), stdout);
                    obuffer_ptr = obuffer;
                    avail_out = buffer_size;
                    continue;
                }
                abort();
            }
        }
    }
    while(1) {
        enum BroccoliResult res = BroccoliConcatFinish(
            &state,
            &avail_out,
            &obuffer_ptr);
        if (res == BroccoliNeedsMoreOutput) {
            fwrite(obuffer, 1, (obuffer_ptr - obuffer), stdout);
            obuffer_ptr = obuffer;
            avail_out = buffer_size;
            continue;
        }
        if (res == BroccoliNeedsMoreInput) {
            abort();
        }
        if (res == BroccoliSuccess) {
            fwrite(obuffer, 1, (obuffer_ptr - obuffer), stdout);
            break;
        }
        abort(); //failure
    }
    return 0;
}

File: src/bin/catbrotli.rs
extern crate brotli;
extern crate core;
use std::env;
use std::fs::File;
use std::io;
use std::io::{Read, Write};
use std::path::Path;

use brotli::concat::{BroCatli, BroCatliResult};
fn usage() {
    writeln!(
        &mut ::std::io::stderr(),
        "Usage: [-w<window_size>] filename0 filename1 filename2..."
    )
    .unwrap();
}
fn read_no_interrupt<R: Read>(r: &mut R, buf: &mut [u8]) -> Result<usize, io::Error> {
    loop {
        match r.read(buf) {
            Err(e) => match e.kind() {
                io::ErrorKind::Interrupted => continue,
                _ => return Err(e),
            },
            Ok(cur_read) => return Ok(cur_read),
        }
    }
}

fn write_no_interrupt<W: Write>(w: &mut W, mut buf: &[u8]) -> Result<usize, io::Error> {
    let mut total_read = 0usize;
    loop {
        match w.write(buf) {
            Err(e) => match e.kind() {
                io::ErrorKind::Interrupted => continue,
                _ => return Err(e),
            },
            Ok(cur_read) => {
                buf = &buf[cur_read..];
                total_read += cur_read;
                if buf.is_empty() {
                    return Ok(total_read);
                }
            }
        }
    }
}

fn main() {
    let mut window_size: Option<u8> = None;
    let mut double_dash = false;
    let mut buffer_size = 4096usize;
    let mut filenames = Vec::<String>::new();
    let mut ostream = io::stdout();
    if env::args_os().len() > 1 {
        for argument in env::args().skip(1) {
            if argument.starts_with("-w") && !double_dash {
                window_size = Some(
                    argument
                        .trim_matches('-')
                        .trim_matches('w')
                        .parse::<i32>()
                        .unwrap() as u8,
                );
                continue;
            }
            if argument.starts_with("-bs") && !double_dash {
                buffer_size = argument
                    .trim_matches('-')
                    .trim_matches('b')
                    .trim_matches('s')
                    .parse::<usize>()
                    .unwrap();
                continue;
            }
            if argument == "--" {
                double_dash = true;
                continue;
            }
            filenames.push(argument);
        }
    } else {
        usage();
        return;
    }
    let mut ibuffer = vec![0u8; buffer_size];
    let mut obuffer = vec![0u8; buffer_size];
    let mut ooffset = 0;
    let mut ioffset;
    let mut bro_cat_li = match window_size {
        Some(ws) => BroCatli::new_with_window_size(ws),
        None => BroCatli::new(),
    };
    for filename in filenames {
        bro_cat_li.new_brotli_file();
        let mut input_file = match File::open(Path::new(&filename)) {
            Err(why) => panic!("couldn't open {:}\n{:}", filename, why),
            Ok(file) => file,
        };
        loop {
            ioffset = 0;
            match read_no_interrupt(&mut input_file, &mut ibuffer[..]) {
                Err(e) => panic!("{}", e),
                Ok(cur_read) => {
                    if cur_read == 0 {
                        break;
                    }
                    loop {
                        match bro_cat_li.stream(
                            &ibuffer[..cur_read],
                            &mut ioffset,
                            &mut obuffer[..],
                            &mut ooffset,
                        ) {
                            BroCatliResult::NeedsMoreOutput => {
                                match write_no_interrupt(&mut ostream, &obuffer[..ooffset]) {
                                    Err(why) => panic!("couldn't write: {:}", why),
                                    Ok(count) => {
                                        assert_eq!(count, ooffset);
                                    }
                                }
                                ooffset = 0;
                            }
                            BroCatliResult::NeedsMoreInput => {
                                break;
                            }
                            BroCatliResult::Success => {
                                panic!("Unexpected state: Success when streaming before finish");
                            }
                            failure => {
                                panic!(
                                    "Failed to concatenate files on {:} {:?}",
                                    filename, failure
                                );
                            }
                        }
                    }
                }
            }
        }
    }
    loop {
        match bro_cat_li.finish(&mut obuffer[..], &mut ooffset) {
            BroCatliResult::NeedsMoreOutput => {
                match write_no_interrupt(&mut ostream, &obuffer[..ooffset]) {
                    Err(why) => panic!("couldn't write: {:}", why),
                    Ok(count) => {
                        assert_eq!(count, ooffset);
                    }
                }
                ooffset = 0;
            }
            BroCatliResult::NeedsMoreInput => {
                panic!("Unexpected EOF");
            }
            BroCatliResult::Success => {
                if ooffset != 0 {
                    match write_no_interrupt(&mut ostream, &obuffer[..ooffset]) {
                        Err(why) => panic!("couldn't write: {:}", why),
                        Ok(count) => {
                            assert_eq!(count, ooffset);
                        }
                    }
                }
                break;
            }
            failure => {
                panic!("{:?}", failure)
            }
        }
    }
}
